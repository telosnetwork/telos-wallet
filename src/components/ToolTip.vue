<script setup lang="ts">
import { ref } from 'vue';
import { directive as vClickAway } from 'vue3-click-away';

import InlineSvg from 'vue-inline-svg';
const infoIcon = require('src/assets/icon--info.svg');
const warningIcon = require('src/assets/icon--warning.svg');

const props = defineProps<{
    text?: string,
    warnings?: string[],
    hideIcon?: boolean,
}>();

let tooltipEnabled = ref(false);

function setTooltipVisibility(enable: boolean) {
    tooltipEnabled.value = enable;
}
</script>

<template>
<div
    v-click-away="() => setTooltipVisibility(false)"
    :class="{
        'c-tooltip': true,
        'c-tooltip--dynamic-size': $slots.default
    }"
    @mouseenter="setTooltipVisibility(true)"
    @mouseleave="setTooltipVisibility(false)"
    @touchend="setTooltipVisibility(!tooltipEnabled)"
>
    <slot></slot>
    <InlineSvg
        v-if="!hideIcon"
        :src="infoIcon"
        class="c-tooltip__icon"
        aria-hidden="true"
    />
    <q-tooltip
        :model-value="tooltipEnabled"
        v-bind="{ ...$attrs }"
        transition-show="scale"
        transition-hide="scale"
        @update:model-value="() => {}"
    >
        <div class="c-tooltip__text-container">
            <div
                v-for="warning in (warnings ?? [])"
                :key="warning"
                class="c-tooltip__warning-container"
            >
                <InlineSvg
                    :src="warningIcon"
                    class="c-tooltip__icon c-tooltip__icon--warning"
                    aria-hidden="true"
                />

                <span class="c-tooltip__text">
                    {{ warning }}
                </span>
            </div>

            <span v-if="text" class="c-tooltip__text">
                {{ text }}
            </span>
        </div>

    </q-tooltip>
</div>
</template>

<style lang="scss">
.c-tooltip {
    display: inline-flex;
    height: 24px;
    width: 24px;
    justify-content: center;
    align-items: center;
    gap: 4px;

    &--dynamic-size {
        width: unset;
        height: unset;
        justify-content: unset;
    }

    &__icon {
        height: 14px;

        path {
            fill: var(--accent-color);
        }

        &--warning {
            path {
                fill: white;
            }
        }
    }

    &__text-container {
        display: flex;
        flex-direction: column;
        gap: 12px;
    }

    &__warning-container {
        display: flex;
        align-items: center;
        gap: 4px;
    }

    &__text {
        @include text--small-bold;
        white-space: pre-line;
    }
}
</style>
