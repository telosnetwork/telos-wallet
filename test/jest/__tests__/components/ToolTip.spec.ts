import { shallowMount } from '@vue/test-utils';
import ToolTip from 'components/ToolTip.vue';
import { stubWithSlot } from 'test/jest/testing-helpers';

describe('ToolTip.vue', () => {
    const globalOpts = {
        stubs: {
            'q-tooltip': stubWithSlot('q-tooltip'),
        },
    };

    it('should render correctly with default options', async () => {
        const wrapper = shallowMount(ToolTip, {
            global: { ...globalOpts },
            props: {
                text: 'Test',
            },
        });

        // no warning text
        expect(wrapper.element).toMatchSnapshot();

        await wrapper.setProps({
            warnings: ['test', 'test2'],
        });

        // with warning text
        expect(wrapper.element).toMatchSnapshot();
    });

    it('should render correctly with hidden icon', () => {
        const wrapper = shallowMount(ToolTip, {
            global: { ...globalOpts },
            props: {
                text: 'Test',
                hideIcon: true,
            },
        });

        // no warning text
        expect(wrapper.element).toMatchSnapshot();
    });

    it('should render correctly with something in its slot', () => {
        const wrapper = shallowMount(ToolTip, {
            global: { ...globalOpts },
            props: {
                text: 'Test',
            },
            slots: {
                default: 'slot content',
            },
        });

        // no warning text
        expect(wrapper.element).toMatchSnapshot();
    });
});
