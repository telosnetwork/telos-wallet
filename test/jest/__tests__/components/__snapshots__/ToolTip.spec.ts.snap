// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolTip.vue should render correctly with default options 1`] = `
<div
  class="c-tooltip"
>
  
  
  <inline-svg-stub
    aria-hidden="true"
    class="c-tooltip__icon"
    keepduringloading="true"
    src=""
    transformsource="[Function]"
  />
  <q-tooltip-stub
    model-value="false"
    transition-hide="scale"
    transition-show="scale"
  >
    
    <div
      class="c-tooltip__text-container"
    >
      
      
      <span
        class="c-tooltip__text"
      >
        Test
      </span>
    </div>
    
  </q-tooltip-stub>
</div>
`;

exports[`ToolTip.vue should render correctly with default options 2`] = `
<div
  class="c-tooltip"
>
  
  
  <inline-svg-stub
    aria-hidden="true"
    class="c-tooltip__icon"
    keepduringloading="true"
    src=""
    transformsource="[Function]"
  />
  <q-tooltip-stub
    model-value="false"
    transition-hide="scale"
    transition-show="scale"
  >
    
    <div
      class="c-tooltip__text-container"
    >
      
      <div
        class="c-tooltip__warning-container"
      >
        <inline-svg-stub
          aria-hidden="true"
          class="c-tooltip__icon c-tooltip__icon--warning"
          keepduringloading="true"
          src=""
          transformsource="[Function]"
        />
        <span
          class="c-tooltip__text"
        >
          test
        </span>
      </div>
      <div
        class="c-tooltip__warning-container"
      >
        <inline-svg-stub
          aria-hidden="true"
          class="c-tooltip__icon c-tooltip__icon--warning"
          keepduringloading="true"
          src=""
          transformsource="[Function]"
        />
        <span
          class="c-tooltip__text"
        >
          test2
        </span>
      </div>
      
      <span
        class="c-tooltip__text"
      >
        Test
      </span>
    </div>
    
  </q-tooltip-stub>
</div>
`;

exports[`ToolTip.vue should render correctly with hidden icon 1`] = `
<div
  class="c-tooltip"
>
  
  
  <!--v-if-->
  <q-tooltip-stub
    model-value="false"
    transition-hide="scale"
    transition-show="scale"
  >
    
    <div
      class="c-tooltip__text-container"
    >
      
      
      <span
        class="c-tooltip__text"
      >
        Test
      </span>
    </div>
    
  </q-tooltip-stub>
</div>
`;

exports[`ToolTip.vue should render correctly with something in its slot 1`] = `
<div
  class="c-tooltip c-tooltip--dynamic-size"
>
  
  slot content
  
  <inline-svg-stub
    aria-hidden="true"
    class="c-tooltip__icon"
    keepduringloading="true"
    src=""
    transformsource="[Function]"
  />
  <q-tooltip-stub
    model-value="false"
    transition-hide="scale"
    transition-show="scale"
  >
    
    <div
      class="c-tooltip__text-container"
    >
      
      
      <span
        class="c-tooltip__text"
      >
        Test
      </span>
    </div>
    
  </q-tooltip-stub>
</div>
`;
